dialer.go

```diff
@@ -109,27 +129,42 @@ func (d *dialer) DialStream(dial DialFN) (Stream, error) {
 		d.lastDialed = now
 	}
 
-	// TODO: support pooling of connections (i.e. keep multiple physical connections in flight)
-	if current == nil || idsExhausted || idled {
-		if current != nil {
+	if uninitialized || idsExhausted || idled {
+		if !uninitialized {
 			// Need to run this on a goroutine because we call back to sessionClosed()
 			// via the session's beforeClose callback, which needs the mutex that's
 			// already locked here.
-			go current.markDefunct()
+			go func() {
+				s, err := getSession(current, sessionAcquisitionTimeout)
+				if err == nil {
+					s.markDefunct()
+				}
+			}()
 		}
 
-		var err error
-		current, err = d.startSession(dial)
-		if err != nil {
-			d.mx.Unlock()
-			return nil, err
-		}
+		log.Debugf("Attempting to start new session")
+		current = eventual.NewValue()
+		d.current = current
+		d.lastDialed = time.Now()
+
+		go d.initSession(current, dial)
 	}
-	id := d.id
-	d.id++
 	d.mx.Unlock()
 
-	c, _ := current.getOrCreateStream(id)
+	_sae, ok := current.GetOrInit(sessionAcquisitionTimeout, retrySessionAfter, func(v eventual.Value) {
+		log.Debugf("Failed to obtain a session within %v, will try to create a new one", retrySessionAfter)
+		d.initSession(v, dial)
```

Comment:
Because we use the new `GetOrInit` function, if this call to `DialStream` hasn't gotten a session (or error) within 5 seconds, it will try to dial a session again.